pkg/scheduler/event_handler.go

```diff
@@ -256,7 +258,66 @@ func (s *Scheduler) updateCluster(_, newObj interface{}) {
 		if features.FeatureGate.Enabled(features.Failover) { // Trigger reschedule on cluster failure only when flag is true.
 			s.enqueueAffectedBinding(newCluster.Name)
 			s.enqueueAffectedClusterBinding(newCluster.Name)
-			return
+		}
+	}
+
+	if meta.IsStatusConditionTrue(newCluster.Status.Conditions, clusterv1alpha1.ClusterConditionReady) {
```

Comment:
I think it should be `!oldCluster.Ready && newCluster.Ready`? Otherwise it would reschedule whenever the cluster is updated

Reply:
Agreed

Reply:
+1

Reply:
Done!

Reply:
https://github.com/karmada-io/karmada/pull/2301#issuecomment-1221963077

Reply:
I think in most cases, the cluster condition would be `ready`, and the status of `Cluster` will constantly be updated (like resource status), which means the relevant RB/CRB will be thrown into queue constantly. That's my concern.

Reply:
@Garrybest what do you think?

Reply:
Agree with @RainbowMango, why don't we only care about label changes here?

Reply:
And when adding cluster, we could add corresponding RB/CRB into queue. But I notice this PR only focuses on cluster updating. What's the intension?

Reply:
> And when adding cluster, we could add corresponding RB/CRB into queue. But I notice this PR only focuses on cluster updating. What's the intension?

Because I think that when adding cluster , cluster may is "NotReady". At this time, rescheduling is meaningless

Reply:
Really? I'm not sure about the cluster establish details. But we should know when the scheduler restarts, we receive the create event instead of update, so I still think it's useful to do this job in adding event. 

Anyway, condition will be transfered into `Taint`, so how about considering these conditions when cluster updating?

1. taints change
2. labels change
3. [cluster fields changes](https://github.com/karmada-io/karmada/blob/822abc540b622bad61f1319cc9bbe7d1363fe102/pkg/util/selector.go#L119-L136)

Reply:
> 1. taints change
> 3. cluster fields changes

Can I use the `Generation` field uniformly to judge?

Reply:
Good idea👍

Reply:
BTW, it would be better to add these conditions into e2e if available😄

Reply:
> Can I use the Generation field uniformly to judge?

I'm afraid not. 
1. We haven't increased the generation by now, I just realized it recently. 
2. For label/annotation change, the generation would not increase.

Reply:
Is there a situation where only the cluster's state changes from `Notready` to `Ready`, while the cluster's labels , taints and cluster fields  haven't changed?



@Garrybest @RainbowMango

Reply:
Yes, probably.  Why do you ask? I don't get your point.

Reply:
<img width="874" alt="image" src="https://user-images.githubusercontent.com/17962021/186327996-dd662820-ab80-4833-a06c-9355cd23a722.png">


The state of the newly added cluster is always changed from `NotReady` to `Ready`. At this time, the label and taint of the cluster will not change. Therefore, if we[ only care about the changes of label and taint in the update event](https://github.com/karmada-io/karmada/pull/2301#discussion_r952331324), we cannot trigger the rescheduling again

Reply:
The initial state of the newly added cluster is `NotReady`, so add  event` cannot trigger rescheduling

Reply:
Now scheduler does not care about whether the cluster is ready or not. Readiness is tranfered as taint. So the e2e test may be not appropriate now?

Reply:
> The initial state of the newly added cluster is NotReady, 

Really? I remember the initial state is `empty`, neither `NotReady` nor`Ready`.
Can you show me the code?

Reply:
I mean, the initial state of the cluster is not ready(empty or NotReady). There is always a change process, and this change process (update event) cannot trigger rescheduling

Reply:
> The initial state of the newly added cluster is NotReady

Again, why do you think that?

On the other hand, we put the binding into the queue in `cluster add` event, correct? Just as @Garrybest said on above, the scheduler doesn't care about the cluster `ready` state now.

Reply:
> we put the binding into the queue in cluster add event

I have added, but now E2E still fails.

<img width="773" alt="image" src="https://user-images.githubusercontent.com/17962021/186616125-d6747bf7-e97e-4b11-b274-25ce9fa0ddd6.png">

Reply:
I know it. 

> The initial state of the newly added cluster is NotReady, so add event cannot trigger rescheduling(the cluster status is not Ready, this cluster will be filtered by the scheduler)

I mean this might not be the root cause of E2E failure, because we are sure the RB can be put into the queue(rescheduling).

Reply:
>  root cause of E2E failure 

I found the reason why E2E failed
https://github.com/karmada-io/karmada/pull/2427

Reply:
Can you explain it in detail?